ONSdigital/census-rh-cucumber
# CR-366 (and CR-394)

# Motivation and Context


This change is required in order to test the happy path scenario for someone using the RHUI as follows:
    Given I am on the RH Start Page
    And I enter valid UAC "aaaabbbbccccdddd" into the text box
    And I click the “Start now” button
    Then I am presented with a page to confirm my address
    And the address confirmation page contains two options
    Given I select the “Yes, this address is correct” option
    And I click the “Continue” button
    Then I am directed to the Census Questionnaire

As part of the above scenario, the UAC "aaaabbbbccccdddd" is added to Firestore, in the relevant gcp project, using the following approach:
    - use the event publisher to send a case_created event, and a uac_updated event, to rabbitmq 
    NB. The uac_updated event contains a hashed version of uac "aaaabbbbccccdddd".
    - wait for rhsvc to receive the events (case_created and uac_updated events), from the relevant queues, and put them in the gcp firestore
    - use the firestoreWait function, from the common test framework, to find out when the data exists in the gcp firestore
    - as soon as the data exists in the gcp firestore then the happy path scenario should be able to put the UAC "aaaabbbbccccdddd" into the census-rh-ui 
    and find that it is accepted as valid.

NB. The above steps require that both rabbitmq and the rhsvc are running at the same time as the scenario is running. The scenario is currently hard-coded to point
to the census-rh-ui url in the gcp dev environment (census-rh-dev project), which is this: http://34.96.85.76/start/. Therefore the rhsvc needs to be configured to
point to census-rh-dev as follows: export GOOGLE_CLOUD_PROJECT="census-rh-dev". The common test framework will use the default gcp project for the environment that it is
running in. Therefore, if the common test framework is run locally, the local environment will need to be set to default to census-rh-dev too: 

gcloud config set project census-rh-dev

Furthermore, this change is required to test the following two unhappy path scenarios:

  @SetUpRH1 @TearDownRH1
  Scenario: Unhappy path - No UAC entered
    Given I am on the RH Start Page
    And I click the “Start now” button
    Then an error message “Enter your access code.” appears

  @SetUpRH1 @TearDownRH1
  Scenario: Unhappy path – Invalid UAC entered
    Given I am on the RH Start Page
    And I enter an invalid UAC "aaaaaaaaaaaaaaaa" into the text box
    And I click the “Start now” button
    Then an error message “Please re-enter your access code and try again.” appears

# What has changed

The happy path, and the two unhappy paths, have been written as the first three scenarios in the following feature file:

/census-rh-cucumber/src/test/resources/integrationtests/RH1.feature

NB. The 'accessTheCensus' feature, which was used for the POC, has been deleted to avoid clashes with RH1. Other feature files are kept as they may be useful 
later on.

The RH1 feature has the following corresponding class, which contains the code for the scenario steps:

/census-rh-cucumber/src/test/java/uk/gov/ons/ctp/integration/rhcucumber/cucSteps/RH1.java

To support the RH1 scenarios, as part of the cucumber/selenium framework, the following page objects have been created and/or modified:
- /census-rh-cucumber/src/test/java/uk/gov/ons/ctp/integration/rhcucumber/selenium/pageobject/Start.java
- /census-rh-cucumber/src/test/java/uk/gov/ons/ctp/integration/rhcucumber/selenium/pageobject/ConfirmAddress.java
- /census-rh-cucumber/src/test/java/uk/gov/ons/ctp/integration/rhcucumber/selenium/pageobject/CensusQuestionnaire.java



# How to test?


In order to run the three RH1 scenarios successfully please follow these steps:
1. Make sure that your local gcp environment is set to census-rh-dev. You can check this using the following command:

gcloud config list --format 'value(core.project)'

And, if it's not currenty set to census-rh-dev, then you can switch to dev with this command:

gcloud config set project census-rh-dev

2. Run rabbitmq locally by moving into the census-rh-service repo, in the terminal, and entering this command:

docker-compose up -d

3. In the same repo then set the GOOGLE_CLOUD_PROJECT environment variable as follows:

export GOOGLE_CLOUD_PROJECT="census-rh-dev"

4. In the same repo then launch census-rh-service as follows:

mvn spring-boot:run

5. Now move into the census-rh-cucumber repo. Run all the cucumber tests, which are currently just the three scenarios in RH1, using the following command:

mvn test
